content/shmem_broadcast.tex

```diff
@@ -72,12 +72,15 @@
     needed to ensure this): The \VAR{pSync} array on all \ac{PE}s in the
     \activeset{} is not still in use from a prior call to a broadcast routine.  The
     \dest{} array on all \ac{PE}s in the \activeset{} is ready to accept the
-    broadcast data.
+    broadcast data. \newtext{The data in the \source{} array on the root \ac{PE} is ready to be broadcasted.}
```

Comment:
> However, this change would introduce a race into the above test since there is no synchronization between the producer PE setting up the source buffer the consumer PEs calling the barrier.

Depending on the implementation, there is **already** a race in this test (line: 118), because (1) there is no explicit checking for the destination buffer to be ready and (2) same destination buffer is used for multiple iterations. The root can overwrite the destination buffers, if root is fast and non-root PEs are slower. So, I don't see, how there is a new race introduced. 


> It's often easy for applications to resolve the data dependence involved with the destination buffer.

Can you please elaborate ?

> In any hierarchical broadcast based on gets, you would still need to synchronize the RAW dependence between the interior nodes in the tree regardless of this semantic. Thus the performance benefit is limited to nodes that are children of the root, which will vanish at large enough scale.

I agree about the benefits at scale. Though this might be valid, why constraint an implementation. 

What bothers me about specifying “The shmem_broadcast cannot start unless the root initiates ..” is that it is a stronger semantic than it is needed for a broadcast, and it prescribes how to implement the broadcast. For broadcast the data on the root's source buffer should be ready, it should not matter whether root initiates or non-root PEs initiate the data transfer.